docs/proposals/improve-status-in-CAPI-resources.md

```diff
@@ -0,0 +1,921 @@
+- Use of the Reason field is required (currently in Cluster API reasons is added only when condition are false)
+- Controllers should apply their conditions to a resource the first time they visit the resource, even if the status is Unknown.
+  (currently Cluster API controllers add conditions at different stages of the reconcile loops)
+- Cluster API is also dropping its own Condition type and start using metav1.Conditions from the Kubernetes API.
+
+The last point have also another implication, which is the removal of the Severity field which is currently used
+to determine priority when merging conditions.
+
+TODO Document how we are going to replace severity (currently prototyping)
+
+### Changes to Machine resource
+
+#### Machine Status
+
+Following changes are implemented to Machine's status:
+
+- Disambiguate usage of ready term by renaming fields used for the provisioning workflow
+- Align to K8s API conventions by deprecating `Phase` and corresponding `LastUpdated`
+- Remove `FailureReason` and `FailureMessage` to get rid of the confusing concept of terminal failures
+- Transition to new, improved, K8s API conventions aligned conditions
+
+Below you can find the relevant fields in Machine Status v1beta2, after v1beta1 removal (end state);
+After golang types, you can find a summary table that also shows how changes will be rolled out according to K8s deprecation rules.
+
+```golang
+type MachineStatus struct {
+
+    // Initialization provides observations of the Machine initialization process.
+    // NOTE: fields in this struct are part of the Cluster API contract and are used to orchestrate initial Machine provisioning.
+    // The value of those fields is never updated after provisioning is completed.
+    // Use conditions to monitor the operational state of the Machine's BootstrapSecret.
+    // +optional
+    Initialization *MachineInitializationStatus `json:"initialization,omitempty"`
+    
+    // Represents the observations of a Machine's current state.
+    // +listType=map
+    // +listMapKey=type
+    Conditions []metav1.Condition `json:"conditions,omitempty"`
+    
+    // Other fields...
+    // NOTE: `Phase`, `LastUpdated`, `FailureReason`, `FailureMessage` fields won't be there anymore
+}
+
+// MachineInitializationStatus provides observations of the Machine initialization process.
+type MachineInitializationStatus struct {
+
+    // BootstrapSecretCreated is true when the bootstrap provider reports that the Machine's boostrap secret is created.
+    // NOTE: this field is part of the Cluster API contract, and it is used to orchestrate initial Machine provisioning.
+    // The value of this field is never updated after provisioning is completed.
+    // Use conditions to monitor the operational state of the Machine's BootstrapSecret.
+    // +optional
+    BootstrapSecretCreated bool `json:"bootstrapSecretCreated"`
```

Comment:
Isn't that hte same as having a phase field though?

Reply:
> Have you considered using an enum for this instead, perhaps it would be better to represent this in a way that allows more states.

This is just a flag to signal signal progress during the inital provisioning.
There will be a `BootstrapConfigReady` condition to surface what is happening before and after BootstrapSecretCreated.

(So the answer is yes, it is just addressed using conditions, like other things having more states)

Reply:
> Isn't that hte same as having a phase field though?

Right yes, I think it is

> This is just a flag to signal signal progress during the inital provisioning.

If you have the condition, and the condition eventually reaches some terminal happy state, and reaching that terminal happy state flicks this bool to true, do we really need the bool, or should the consumer of this field just check the condition status?

Reply:
> The value of this field is never updated after provisioning is completed.
   Use conditions to monitor the operational state of the Machine's BootstrapSecret.
   
The condition can flick after initial provisioning is completed (the condition doesn't have a "terminal" happy state).
e.g. After initial provisioning of the secret, the condition goes to false if the controller fails to renew the token in the workload cluster.

Reply:
If the condition flicks to false, why would this field not also flick to false? Is the semantic of the field here that the secret was once available rather than is currently available? What use case are we solving by having that separate?

Reply:
I think we're coming from that we want to continue to support the "initial provisioning / bootstraping" workflow of Machines. We are only using the BootstrapSecret *once* for a **Machine** during initial bootstrap.

From the other side we see value in having continuous Ready state for the InfraMachine objects, so that becomes then the question if we should do the same for BootstrapConfig. Also there are cases where BootstrapConfigs are used more than once (at least MachinePools).

On the other side the `BootstrapConfigReady` condition can give us precise error information in case the condition doesn't become ready, which is not really possible with a bool or enum field.

Reply:
FYI I have added a new non-goal/paragraph about the impacts this proposal on the Cluster API contract; hopefully this could also help to address this discussion (changes are in a separated commit for better review)

TL;DR
This proposal is not going to change _how_ the Cluster API contract with infrastructure, bootstrap and control providers currently works (by using status fields like `status.ready`, bool).

However, this proposal going to change a few field names in the Cluster API contract with the only goal _to disambiguate the usage of the ready term_ by renaming fields used for the initial provisioning workflow (e.g. from `status.ready` to `status.initialization.dataSecretCreated`, which surfaces as `status.initialization.bootstapDataSecretCreated`).

While doing this, we also try to better document what is the current behaviour of Cluster API contract fields.

Changing the current behaviour of Cluster API contract fields, e.g allow those fields to flick value after provisioning, or replace them with conditions is out of scope of this proposal (note: I'm not saying it is a good or bad thing, I'm just trying to keep the scope of this work under control 😅)